onnxruntime/core/providers/qnn/builder/opbuilder/reduce_op_builder.cc

```diff
@@ -0,0 +1,346 @@
+                  "Expect QNN Reduce* operator to have at most rank(input[0]) axes elements.");
+      reduce_axes.resize(axes_bytes.size() / sizeof(AxesOnnxIntType));
+
+      auto src_span = gsl::make_span(axes_bytes.data(), axes_bytes.size());
+      auto dst_span = gsl::make_span(reduce_axes.data(), reduce_axes.size());
+
+      // Copy initializer bytes (stored in little-endian order) to vector of int64_t.
+      // ReadLittleEndian returns a status error if the source and destination spans do not have
+      // matching byte sizes.
+      ORT_RETURN_IF_ERROR(onnxruntime::utils::ReadLittleEndian(src_span, dst_span));
+    }
+  }
+
+  if (reduce_axes.size() == 0) {
+    // Use default axes of (0, 1, 2, ..., input_rank - 1)
+    for (size_t i = 0; i < input_rank; ++i) {
+      axes_set.insert(static_cast<AxesOnnxIntType>(i));
+    }
+  } else {
+    // QNN does not support negative axes values. Fix negative values by adding the input rank.
+    for (auto ax : reduce_axes) {
+      AxesOnnxIntType positive_axis = (ax < 0) ? (ax + static_cast<AxesOnnxIntType>(input_rank)) : ax;
+      axes_set.insert(positive_axis);
+    }
+  }
+
+  if (axes_set.size() > input_rank) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
+                           "QNN EP: \"axes\" input must have shape [M] where 0 < M <= rank(input[0])");
+  }
+
+  return Status::OK();
+}
+
+Status ReduceOpBuilder::IsOpSupported(QnnModelWrapper& qnn_model_wrapper,
+                                      const NodeUnit& node_unit,
+                                      const logging::Logger& logger,
+                                      bool is_quantized_model) const {
+  ORT_UNUSED_PARAMETER(logger);
+  ReduceOpType reduce_op_type = GetReduceOpType(node_unit.OpType());
+  if (reduce_op_type == ReduceOpType::REDUCE_OP_TYPE_UNKNOWN) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN EP: Unknown reduce operator ", node_unit.OpType());
+  }
+
+  if (reduce_op_type == ReduceOpType::REDUCE_OP_TYPE_PROD && is_quantized_model) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN EP: ReduceProd operator not supported by HTP backend.");
+  }
+
+  NodeAttrHelper node_attr_helper(node_unit);
+  const auto& inputs = node_unit.Inputs();
+  const auto& data_input = inputs[0];
+  std::vector<uint32_t> input_data_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(data_input.node_arg, input_data_shape),
+                    "Cannot get shape of data input");
+  const size_t input_data_rank = input_data_shape.size();
+
+  // Check allowed data input type.
+  const std::vector<const char*>& allowed_data_types = is_quantized_model ? data_input_quant_types[reduce_op_type] : data_input_types[reduce_op_type];
+  ONNX_NAMESPACE::DataType input_data_type = data_input.node_arg.Type();
+  ORT_RETURN_IF_ERROR(IsTypeAllowed(input_data_type, allowed_data_types,
+                                    "Invalid data input type for QNN reduce operator."));
```

Comment:
I'm not sure whether this is the right approach. My concern is we need to maintain a supported type list here, which means we need to keep it updated with new Qnn versions once they have new types supported. Actually UFIXED_POINT_16 is also supported here by HTP backend.
The existing approach is leave it to Qnn Op validation API to do the job. We just create the Qnn op with what we get from Onnx node. It should fails Qnn Op validation API if the given data type is not supported.

Reply:
Removed this check